infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/fsm/sink/SinkVerifyRemoteLeader.java

```diff
@@ -0,0 +1,210 @@
+                    pendingLeadershipQueries.put(nodeId, leadershipRequestCf);
+                }
+
+                // Block until all leadership requests are completed, or a leader is discovered.
+                while (pendingLeadershipQueries.size() != 0) {
+                    LogReplication.LogReplicationLeadershipResponseMsg leadershipResponse = (LogReplication.LogReplicationLeadershipResponseMsg)
+                            CompletableFuture.anyOf(pendingLeadershipQueries.values()
+                                    .toArray(new CompletableFuture<?>[pendingLeadershipQueries.size()])).get(CorfuLogReplicationRuntime.DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS);
+
+                    if (leadershipResponse.getIsLeader()) {
+                        log.info("Received Leadership Response :: leader for remote cluster, node={}", leadershipResponse.getNodeId());
+                        leader = leadershipResponse.getNodeId();
+                        setRemoteLeaderNodeId(leader);
+
+                        // Remove all CF, based on the assumption that one leader response is the expectation.
+                        pendingLeadershipQueries.clear();
+
+                        // A new leader has been found,
+                        input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_FOUND, leader));
+                        log.debug("Exit :: leadership verification");
+                        return;
+                    } else {
+                        log.debug("Received Leadership Response :: node {} is not the leader", leadershipResponse.getNodeId());
+
+                        // Remove CF for completed request
+                        pendingLeadershipQueries.remove(leadershipResponse.getNodeId());
+                    }
+                }
+
+                // No remote leader was found, retry leadership
+                input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_NOT_FOUND, leader));
+
+            } catch (Exception ex) {
+                log.warn("Exception caught while verifying remote leader.", ex);
+                input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_NOT_FOUND, leader));
+            }
+        } else {
+            log.info("Remote Leader already present {}. Skip leader verification.", getRemoteLeaderNodeId().get());
+        }
+
+        log.debug("Exit :: leadership verification");
+    }
+
+    private void subscribeAndStartReplication() {
+        CorfuMessage.ResponsePayloadMsg payload =
+                CorfuMessage.ResponsePayloadMsg.newBuilder().setLrSubscribeRequest(
+                        LogReplication.SubscribeToReplicationMsg.newBuilder().build()).build();
+
+        log.info("Sending the subscribe msg {}", payload);
+        router.sendResponse(payload, getRemoteLeaderNodeId().get());
+    }
```

Comment:
From debugging ITs I felt that it might be useful to log the session that is sending subscribe request, as the only thing matters in payload part is PayloadCase that is already suggested in log msg